frontend/test/metabase/lib/expressions/generator.js

```diff
@@ -167,11 +95,411 @@ export function generateExpression(seed) {
   };
 
-  let depth = 17;
+  function generateTokenSeq() {
+    const literal = () => {
+      const exp = () => randomItem(["", "-", "+"]) + randomInt(1e2);
+      const number = () => oneOf([zero, one, integer, float1, float2])();
+      const sci = () => number() + randomItem(["e", "E"]) + exp();
+      return {
+        type: NODE.Literal,
+        value: oneOf([number, sci, string])(),
+      };
+    };
+
+    const field = () => {
+      const fk = () => "[" + identifier() + " → " + identifier() + "]";
+      const bracketedName = () => "[" + identifier() + "]";
+      const name = oneOf([identifier, fk, bracketedName])();
+      return {
+        type: NODE.Field,
+        value: name,
+      };
+    };
+
+    const unary = () => {
+      return {
+        type: NODE.Unary,
+        op: randomItem(["-", "NOT "]),
+        child: expression(),
+      };
+    };
+
+    const binary = () => {
+      return {
+        type: NODE.Binary,
+        op: randomItem([
+          "+",
+          "-",
+          "*",
+          "/",
+          "=",
+          "!=",
+          "<",
+          ">",
+          "<=",
+          ">=",
+          " AND ",
+          " OR ",
+        ]),
+        left: expression(),
+        right: expression(),
+      };
+    };
+
+    const call = () => {
+      const count = randomInt(5);
+      return {
+        type: NODE.FunctionCall,
+        value: identifier(),
+        params: listOf(count, [expression])(),
+      };
+    };
+
+    const group = () => {
+      return {
+        type: NODE.Group,
+        child: primary(),
+      };
+    };
+
+    const primary = () => {
+      --depth;
+      const node = oneOf([field, literal, unary, binary, call, group])();
+      ++depth;
+      return node;
+    };
+    const expression = () => (depth <= 0 ? literal() : primary());
+
+    const format = node => {
+      const spaces = () => listOf(1, [space, () => ""])().join("");
+      const blank = ch => spaces() + ch + spaces();
+      let str = null;
+      const { type, value, op, left, right, child, params } = node;
+      switch (type) {
+        case NODE.Field:
+        case NODE.Literal:
+          str = value;
+          break;
+        case NODE.Unary:
+          str = blank(op) + format(child);
+          break;
+        case NODE.Binary:
+          str = format(left) + blank(op) + format(right);
+          break;
+        case NODE.FunctionCall:
+          str = value + blank("(") + params.map(format).join(", ") + blank(")");
+          break;
+        case NODE.Group:
+          str = blank("(") + format(child) + blank(")");
+          break;
+      }
+
+      if (str === null) {
+        throw new Error(`Unknown AST node ${type}`);
+      }
+      return String(str);
+    };
+
+    let depth = 17;
+
+    const tree = expression();
+    return { tree, expression: format(tree) };
+  }
+
+  function generateValidSyntax(type) {
+    // LIMITATION: known issue in the current Chevrotain-based parser
```

Comment:
How are we handling these limitations in the new parser? Do we want to manually add tests describe their behavior?

Reply:
I like that suggestion. While discovering each limitation (and I don't think I know all of them already), I'll collect them (in a separate PR) and turn them into unit tests with `it.skip`. They need to be unskipped once a better parser is in place.